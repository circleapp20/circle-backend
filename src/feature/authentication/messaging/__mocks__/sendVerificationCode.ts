export const sendVerificationCodeByEmail = jest.fn();
export const sendVerificationCodeBySMS = jest.fn();
export const sendVerificationCodeByMedia = jest.fn();
