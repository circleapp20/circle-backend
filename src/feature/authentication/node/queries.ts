import { Users } from 'shared/common/schema/users';
import { EntityManager } from 'typeorm';

export const countMatchingIdAndCodeQuery = (
	manager: EntityManager,
	id: string,
	verificationCode: string
) => {
	return manager
		.createQueryBuilder(Users, 'u')
		.where('u.id = :id', { id })
		.andWhere('u.verificationCode = :code', { code: verificationCode })
		.getCount();
};

export const updateUserVerificationCodeQuery = (
	manager: EntityManager,
	values: { id: string; verificationCode: string }
) => {
	const { id, verificationCode } = values;
	return manager
		.createQueryBuilder(Users, 'u')
		.update()
		.set({ verificationCode })
		.where('u.id = :id', { id })
		.execute();
};
