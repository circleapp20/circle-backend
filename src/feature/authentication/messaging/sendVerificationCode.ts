import { printToConsole } from 'base/common/utilities';
import { getMailHeaders, sendMail, sendSMSMessage } from 'base/node/messaging';
import { getVerificationEmailText } from 'feature/authentication/common/strings';

export const sendVerificationCodeByEmail = (code: string, email: string) => {
	if (!email) return;
	const message = getVerificationEmailText(code);
	const emailHeaders = getMailHeaders({
		to: email,
		subject: 'Circle Verification Code',
		html: message
	});
	sendMail(emailHeaders).catch(printToConsole);
};

export const sendVerificationCodeBySMS = (code: string, phoneNumber: string) => {
	if (!phoneNumber) return;
	const message = `<#> Your circle verification code is :${code}\njdzOJsciqVb`;
	sendSMSMessage({ message, to: phoneNumber }).catch(printToConsole);
};

export const sendVerificationCodeByMedia = (args: {
	verificationCode: string;
	media: 'email' | 'phoneNumber';
	email?: string;
	phoneNumber?: string;
}) => {
	const { verificationCode, media, email, phoneNumber } = args;
	if (email && media === 'email') return sendVerificationCodeByEmail(verificationCode, email);
	if (phoneNumber && media === 'phoneNumber') {
		return sendVerificationCodeBySMS(verificationCode, phoneNumber);
	}
	return;
};
