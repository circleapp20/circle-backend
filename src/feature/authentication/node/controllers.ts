import { Response } from 'express';
import {
	checkUserVerificationCode,
	createUserProfileWithDefaultValues,
	getUserAccountWithCredentials,
	getUserProfileById,
	verifyUserLoginCredentials
} from 'feature/authentication/node/dataService';
import { sendVerificationCodeByMedia } from 'feature/authentication/node/messaging';
import { getResponseData } from 'shared/common/utilities';
import { Constants } from 'shared/constants';
import { IAuthUser, IRequest } from 'shared/types';

export const verifyUserCredentials = async (req: IRequest, res: Response) => {
	const user = await createUserProfileWithDefaultValues(req.body.data);

	sendVerificationCodeByMedia({
		media: user.email ? 'email' : 'phoneNumber',
		verificationCode: user.verificationCode,
		email: user.email,
		phoneNumber: user.phoneNumber
	});

	const responseData = getResponseData(user);
	res.status(Constants.status.CREATED).json(responseData);
};

export const verifyUserVerificationCode = async (req: IRequest, res: Response) => {
	const { id }: IAuthUser = req.user;
	const data = Object.assign({}, req.body.data, { id });
	const status = await checkUserVerificationCode(data);
	const responseData = getResponseData(status);
	res.status(Constants.status.CREATED).json(responseData);
};

export const verifyUserLogin = async (req: IRequest, res: Response) => {
	const user = await verifyUserLoginCredentials(req.body.data);
	const responseData = getResponseData(user);
	res.status(Constants.status.CREATED).json(responseData);
};

export const resendUserVerificationCode = async (req: IRequest, res: Response) => {
	const { media }: any = req.query;
	const user = await getUserProfileById(req.user.id);

	sendVerificationCodeByMedia({
		media: media ? media : user.email ? 'email' : 'phoneNumber',
		verificationCode: user.verificationCode,
		email: user.email,
		phoneNumber: user.phoneNumber
	});

	const responseData = getResponseData(true);
	res.status(Constants.status.CREATED).json(responseData);
};

/**
 * the controller verifies the user's credentials in the database
 * and sends the verification code to the user by email or password
 * should the user have both email and password, a response of message
 * is return to the user to request which media to send the verification
 * code to either email or phone number
 */
export const verifyUserCredentialsForPasswordReset = async (req: IRequest, res: Response) => {
	const profile = await getUserAccountWithCredentials(req.body.data);

	const { email, verificationCode, phoneNumber } = profile;
	let message = 'Verification code cannot be sent. User has both email and phone number';

	sendVerificationCodeByMedia({
		media: phoneNumber ? 'phoneNumber' : 'email',
		verificationCode,
		email: email && !phoneNumber ? email : '',
		phoneNumber: !email && phoneNumber ? phoneNumber : ''
	});

	if (email && !phoneNumber) {
		message = "Verification code sent to user's email";
	} else if (!email && phoneNumber) {
		message = "Verification code sent to user's phone number";
	}

	const responseData = getResponseData({ user: profile, message });
	res.status(Constants.status.CREATED).json(responseData);
};
