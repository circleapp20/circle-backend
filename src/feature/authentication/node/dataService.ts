import bcryptjs from 'bcryptjs';
import {
	countMatchingIdAndCodeQuery,
	updateUserVerificationCodeQuery
} from 'feature/authentication/node/queries';
import { getBadRequestError } from 'shared/common/errors';
import { generateCodeFromNumber } from 'shared/common/utilities';
import { Constants } from 'shared/constants';
import { runInsertQuery, runInTransaction, runQuery } from 'shared/node/database';
import {
	addUserProfileQuery,
	getUserByCredentialsQuery,
	getUserByIdQuery
} from 'shared/node/queries';
import { getSignedAuthToken } from 'shared/node/validation';
import { IAddUserProfile } from 'shared/types';
import { EntityManager } from 'typeorm';

export const addUserTransaction = (email = '', phoneNumber = '') => {
	return async (manager: EntityManager) => {
		// generate a verification code for the user
		const verificationCode = generateCodeFromNumber();

		// construct a default user info
		const profile: IAddUserProfile = {
			username: '',
			password: '',
			dob: new Date(),
			image: '',
			biography: '',
			email: email || '',
			phoneNumber: phoneNumber || '',
			isEmailVerified: false,
			verificationCode,
			roles: [Constants.privileges.USER],
			name: ''
		};

		const [user] = await runInsertQuery(addUserProfileQuery, [profile], manager);

		return runQuery(getUserByIdQuery, [user.id], manager);
	};
};

export const createUserProfileWithDefaultValues = async (data: {
	email?: string;
	phoneNumber?: string;
}) => {
	const user = await runQuery(getUserByCredentialsQuery, [data]);
	if (user) throw getBadRequestError('User already exists');

	// save user profile into the database
	const userProfile = await runInTransaction(addUserTransaction(data.email, data.phoneNumber));

	// add token to user profile
	const token = getSignedAuthToken({ id: userProfile!.id, roles: [Constants.privileges.USER] });
	const profile = Object.assign({}, userProfile, { token });

	return profile;
};

export const checkUserVerificationCode = async (data: { id: string; verificationCode: string }) => {
	const { id, verificationCode } = data;

	if (!id || !verificationCode) {
		throw getBadRequestError('user id and verification code are required');
	}

	const count = await runQuery(countMatchingIdAndCodeQuery, [id, verificationCode]);
	if (!count) throw getBadRequestError('invalid verification code');

	return true;
};

export const verifyUserLoginCredentials = async (data: {
	email?: string;
	phoneNumber?: string;
	username?: string;
	password: string;
}) => {
	const { phoneNumber, password, email, username } = data;
	if (!phoneNumber && !email && !username) {
		throw getBadRequestError('phoneNumber, email or username is required');
	}
	if (!password) throw getBadRequestError('password is required');

	const user = await runQuery(getUserByCredentialsQuery, [data]);
	if (!user) throw getBadRequestError('Invalid user account');

	if (!bcryptjs.compareSync(password, user.password)) {
		throw getBadRequestError('Invalid user account');
	}

	const { password: hashedPassword, ...other } = user;

	// create user token
	const token = getSignedAuthToken({ id: user.id, roles: user.roles });
	const profile = Object.assign({}, other, { token });

	return profile;
};

export const getUserProfileById = async (id: string) => {
	const user = await runQuery(getUserByIdQuery, [id]);
	if (!user) throw getBadRequestError('Invalid user account');
	return user;
};

/**
 * the service function validates the user's credentials in the
 * database to check if the user exists
 * if the user does exists, a new verification code is generated
 * and stored in the database
 */
export const getUserAccountWithCredentials = async (args: {
	email?: string;
	phoneNumber?: string;
	username?: string;
}) => {
	const user = await runQuery(getUserByCredentialsQuery, [args]);
	if (!user) throw getBadRequestError('Invalid user account');

	// generate new verification code
	const verificationCode = generateCodeFromNumber();

	// update user profile
	await runQuery(updateUserVerificationCodeQuery, [{ id: user.id, verificationCode }]);

	const { password, ...other } = user;

	const token = getSignedAuthToken({ id: user.id, roles: user.roles });
	const profile = Object.assign({}, other, { token, verificationCode });

	return profile;
};
