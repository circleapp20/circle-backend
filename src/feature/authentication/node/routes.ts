import * as schema from 'feature/authentication/common/dataSchema';
import * as controller from 'feature/authentication/node/controllers';
import { Constants } from 'shared/constants';
import { IApiRoute } from 'shared/types';

export const authRoutes: IApiRoute[] = [
	{
		path: '/auth/sign-up',
		method: 'post',
		controller: controller.verifyUserCredentials,
		schema: schema.VerifyUserCredentialsSchema
	},
	{
		path: '/auth/code',
		method: 'post',
		controller: controller.verifyUserVerificationCode,
		privileges: [Constants.privileges.SUPER_ADMIN, Constants.privileges.USER],
		schema: schema.VerifyVerificationCodeSchema
	},
	{
		path: '/auth/sign-in',
		method: 'post',
		controller: controller.verifyUserLogin,
		schema: schema.VerifyUserLoginSchema
	},
	{
		path: '/auth/code/resend',
		method: 'post',
		controller: controller.resendUserVerificationCode,
		privileges: [Constants.privileges.USER]
	},
	{
		path: '/auth/credentials',
		method: 'post',
		controller: controller.verifyUserCredentialsForPasswordReset
	}
];
