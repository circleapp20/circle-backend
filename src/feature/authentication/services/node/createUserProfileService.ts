import { SERVER } from 'base/config/server';
import { IAddUserProfile } from 'base/types';
import { getBadRequestError } from 'base/utils/errors/node/badRequestError';
import { generateCodeFromNumber } from 'base/utils/node/codeGenerator';
import { getSignedAuthToken } from 'base/utils/server/validation';
import { runInsertQuery, runInTransaction, runQuery } from 'core/database/queryRunners';
import { decryptData, encryptData } from 'core/encryption/node/encryption';
import {
	addUserProfileQuery,
	getUserByCredentialsQuery,
	getUserByIdQuery
} from 'core/queries/userQueries';
import { EntityManager } from 'typeorm';

export const addUserTransaction = (email = '', phoneNumber = '') => {
	return async (manager: EntityManager) => {
		// generate a verification code for the user
		const verificationCode = generateCodeFromNumber();
		const encryptedVerificationCode = encryptData({ text: verificationCode });

		// construct a default user info
		const profile: IAddUserProfile = {
			username: '',
			password: '',
			dob: new Date(),
			image: '',
			biography: '',
			email: email || '',
			phoneNumber: phoneNumber || '',
			isEmailVerified: false,
			verificationCode: encryptedVerificationCode,
			roles: [SERVER.privileges.USER],
			name: ''
		};

		const [user] = await runInsertQuery(addUserProfileQuery, [profile], manager);

		return runQuery(getUserByIdQuery, [user.id], manager);
	};
};

export const createUserProfileWithDefaultValues = async (data: {
	email?: string;
	phoneNumber?: string;
}) => {
	const user = await runQuery(getUserByCredentialsQuery, [data]);
	if (user) throw getBadRequestError('User already exists');

	// save user profile into the database
	const userProfile = await runInTransaction(addUserTransaction(data.email, data.phoneNumber));
	const verificationCode = decryptData({ encryptedText: userProfile!.verificationCode });

	// add token to user profile
	const token = getSignedAuthToken({
		id: userProfile!.id,
		roles: [SERVER.privileges.USER]
	});
	const profile = Object.assign({}, userProfile, { token, verificationCode });

	return profile;
};
