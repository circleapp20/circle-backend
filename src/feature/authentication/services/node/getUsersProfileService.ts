import { getBadRequestError } from 'base/common/errors';
import * as sharedQueries from 'base/node/queries';
import { runQuery } from 'core/node/database/queryRunners';
import { decryptData } from 'core/node/encryption';

export const getUserProfileById = async (id: string) => {
	const user = await runQuery(sharedQueries.getUserByIdQuery, [id]);
	if (!user) throw getBadRequestError('Invalid user account');
	const { password, verificationCode, ...other } = user;
	const decryptedVerificationCode = decryptData({ encryptedText: verificationCode });
	return Object.assign({}, other, { verificationCode: decryptedVerificationCode });
};
