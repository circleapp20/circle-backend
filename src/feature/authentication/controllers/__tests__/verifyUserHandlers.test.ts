import { SERVER } from 'base/config/server';
import { entityManager } from 'base/testUtils/node/entityManager';
import bcryptjs from 'bcryptjs';
import { encryptData } from 'core/encryption/node/encryption';
import faker from 'faker';
import {
	verifyUserCredentials,
	verifyUserCredentialsForPasswordReset,
	verifyUserLogin,
	verifyUserVerificationCode
} from 'feature/authentication/controllers/verifyUserHandlers';
import { sendVerificationCodeByMedia } from 'feature/authentication/messaging/sendVerificationCode';
import { createUserFixture } from 'fixtures/users';

jest.mock('core/database/queryRunners');
jest.mock('core/models/node/users');
jest.mock('core/models/node/fellows');
jest.mock('core/models/node/locations');
jest.mock('feature/authentication/messaging/sendVerificationCode');

beforeEach(() => jest.clearAllMocks());

const responseMock: any = {
	status: jest.fn().mockReturnThis(),
	json: jest.fn().mockReturnThis()
};

describe('#verifyUserCredentials', () => {
	const requestMock: any = { body: { data: { email: 'test@test.com' } } };
	const vCode = faker.random.word();
	const profile = {
		...createUserFixture(),
		verificationCode: encryptData({ text: vCode })
	};

	test('should create a new user', async () => {
		entityManager.getOne.mockReturnValueOnce(null).mockReturnValueOnce(profile);
		entityManager.execute.mockReturnValueOnce({ generateMaps: [profile] });
		await verifyUserCredentials(requestMock, responseMock);
		expect(responseMock.status).toHaveBeenCalledWith(SERVER.status.CREATED);
	});

	test('should not create a user if exists', (done) => {
		entityManager.getOne.mockReturnValueOnce(profile);
		verifyUserCredentials(requestMock, responseMock).catch(({ message }) => {
			expect(responseMock.status).not.toHaveBeenCalled();
			expect(message).toBe('User already exists');
			done();
		});
	});

	test('should send verification by email', async () => {
		entityManager.getOne.mockReturnValueOnce(null).mockReturnValueOnce(profile);
		entityManager.execute.mockReturnValueOnce({ generateMaps: [profile] });
		await verifyUserCredentials(requestMock, responseMock);
		expect(sendVerificationCodeByMedia).toHaveBeenCalledWith(
			expect.objectContaining({
				media: expect.stringMatching('email'),
				verificationCode: vCode,
				email: expect.stringMatching(profile.email)
			})
		);
	});

	test('should send verification code by sms if phoneNumber', async () => {
		const data = { ...profile, email: '' };
		entityManager.getOne.mockReturnValueOnce(null).mockReturnValueOnce(data);
		entityManager.execute.mockReturnValueOnce({ generateMaps: [data] });
		const req: any = { body: { data: { phoneNumber: data.phoneNumber } } };
		await verifyUserCredentials(req, responseMock);
		expect(sendVerificationCodeByMedia).toHaveBeenCalledWith(
			expect.objectContaining({
				media: expect.stringMatching('phoneNumber'),
				verificationCode: vCode,
				phoneNumber: expect.any(String)
			})
		);
	});
});

describe('#verifyUserVerificationCode', () => {
	const profile = createUserFixture();

	const requestMock: any = {
		body: { data: { verificationCode: '123456' } },
		user: { id: profile.id }
	};
	const user = { ...profile, verificationCode: encryptData({ text: '123456' }) };

	test('should send a status of 201 with the json data', async () => {
		entityManager.getOne.mockReturnValueOnce(user);
		await verifyUserVerificationCode(requestMock, responseMock);
		expect(responseMock.status).toHaveBeenCalledWith(SERVER.status.CREATED);
		expect(responseMock.json).toHaveBeenCalledWith({
			data: true,
			success: true
		});
	});

	test('should throw when user is invalid', () => {
		entityManager.getOne.mockReturnValueOnce(null);
		expect(
			verifyUserVerificationCode(requestMock, responseMock)
		).rejects.toThrowErrorMatchingSnapshot();
	});

	test('should throw when verification code is invalid', () => {
		entityManager.getOne.mockReturnValueOnce(user);
		requestMock.body.data.verificationCode = '098765';
		expect(
			verifyUserVerificationCode(requestMock, responseMock)
		).rejects.toThrowErrorMatchingSnapshot();
	});
});

describe('#verifyUserLogin', () => {
	const user = createUserFixture();
	const req: any = { body: { data: { email: user.email, password: user.password } } };

	test('should send a status of 201 for successful login', async () => {
		entityManager.getOne.mockReturnValueOnce({
			...user,
			password: bcryptjs.hashSync(user.password)
		});
		await verifyUserLogin(req, responseMock);
		expect(responseMock.status).toHaveBeenCalledWith(SERVER.status.CREATED);
	});
});

describe('#verifyUserCredentialsForPasswordReset', () => {
	const profile = { ...createUserFixture(), verificationCode: encryptData({ text: '123456' }) };
	const req: any = { body: { data: { email: profile.email } } };

	test('should send verification code to user email', async () => {
		entityManager.getOne.mockReturnValueOnce({ ...profile, phoneNumber: '' });
		await verifyUserCredentialsForPasswordReset(req, responseMock);
		expect(sendVerificationCodeByMedia).toHaveBeenCalledWith(
			expect.objectContaining({ media: 'email' })
		);
	});

	test('should send verification code to user phone number', async () => {
		entityManager.getOne.mockReturnValueOnce({ ...profile, email: '' });
		await verifyUserCredentialsForPasswordReset(req, responseMock);
		expect(sendVerificationCodeByMedia).toHaveBeenCalledWith(
			expect.objectContaining({
				media: expect.stringMatching('phoneNumber')
			})
		);
	});

	test('should not send verification code if user has both phone number and email', async () => {
		entityManager.getOne.mockReturnValueOnce(profile);
		await verifyUserCredentialsForPasswordReset(req, responseMock);
		expect(responseMock.json).toHaveBeenCalledWith(
			expect.objectContaining({
				data: expect.objectContaining({
					message: expect.stringContaining(
						'Verification code cannot be sent. User has both email and phone number'
					)
				}),
				success: expect.any(Boolean)
			})
		);
	});

	test('should send status of 201', async () => {
		entityManager.getOne.mockReturnValueOnce(profile);
		await verifyUserCredentialsForPasswordReset(req, responseMock);
		expect(responseMock.status).toHaveBeenCalledWith(SERVER.status.CREATED);
	});
});
