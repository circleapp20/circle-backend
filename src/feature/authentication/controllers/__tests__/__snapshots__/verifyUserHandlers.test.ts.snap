// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#verifyUserVerificationCode should throw when user is invalid 1`] = `"Invalid user account"`;

exports[`#verifyUserVerificationCode should throw when verification code is invalid 1`] = `"Invalid verification code"`;
